torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -457,20 +457,21 @@ def _broadcast_params(self) -> None:
                 # Direct broadcasts only
                 for param in params:
                     if not self.should_bucket_param[i_param]:
-                        self.work_handles.append(
-                            dist.broadcast(tensor=param.data, src=global_src_rank, group=self.group, async_op=True)
+                        last_work_handle = dist.broadcast(
+                            tensor=param.data, src=global_src_rank, group=self.group, async_op=True
                         )
+
                     i_param += 1
 
                 # Bucket broadcasts
-                self.work_handles.append(
-                    dist.broadcast(tensor=bucket, src=global_src_rank, group=self.group, async_op=True)
-                )
-
-        # Consume all async calls
-        while len(self.work_handles) > 0:
-            work_handle = self.work_handles.popleft()
-            work_handle.wait()
+                if bucket.numel() > 0:
+                    last_work_handle = dist.broadcast(
+                        tensor=bucket, src=global_src_rank, group=self.group, async_op=True
+                    )
+
+        # Wait for the last call in line
+        if last_work_handle is not None:
+            last_work_handle.wait()
```

Comment:
are we confident that all c10d backend will enforce ordering on collective completion?

Reply:
that's what I've always been told explicitly and coherent with the cuda per stream guarantee, but not sure about the other backends now that you ask, good point